# Compute performance metrics for the given Y and Y_score without sklearn

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import warnings 
warnings.filterwarnings("ignore")



Bad key "text.kerning_factor" on line 4 in
C:\Users\utsav\Anaconda3\lib\site-packages\matplotlib\mpl-data\stylelib\_classic_test_patch.mplstyle.
You probably need to get an updated matplotlibrc file from
http://github.com/matplotlib/matplotlib/blob/master/matplotlibrc.template
or from the matplotlib source distribution


<pre>
<font color='red'><b>A.</b></font> Compute performance metrics for the given data <strong>5_a.csv</strong>
   <b>Note 1:</b> in this data you can see number of positive points >> number of negatives points
   <b>Note 2:</b> use pandas or numpy to read the data from <b>5_a.csv</b>
   <b>Note 3:</b> you need to derive the class labels from given score</pre> $y^{pred}= \text{[0 if y_score < 0.5 else 1]}$

<pre>
<ol>
<li> Compute Confusion Matrix </li>
<li> Compute F1 Score </li>
<li> Compute AUC Score, you need to compute different thresholds and for each threshold compute tpr,fpr and then use               numpy.trapz(tpr_array, fpr_array) <a href='https://stackoverflow.com/q/53603376/4084039'>https://stackoverflow.com/q/53603376/4084039</a>, <a href='https://stackoverflow.com/a/39678975/4084039'>https://stackoverflow.com/a/39678975/4084039</a> Note: it should be numpy.trapz(tpr_array, fpr_array) not numpy.trapz(fpr_array, tpr_array)</li>
<li> Compute Accuracy Score </li>
</ol>
</pre>

In [2]:
df=pd.read_csv('5_a.csv')
df.head(5)


,y,proba
0,1.0,0.637387
1,1.0,0.635165
2,1.0,0.766586
3,1.0,0.724564
4,1.0,0.889199


In [3]:
def add_column(df):
    df['Y_pred'] = df['proba'].apply(lambda x: 1 if x > 0.5 else 0)
    return df



# Task 1 for 5_a.csv

# Calculate Confusion Matrix

In [4]:
def confusion_matrix(df):
    True_Positive= ((df['y']==1) & (df['Y_pred']==1)).sum()
            
    True_Negative = ((df['y']==0) & (df['Y_pred']==0)).sum()
            
    False_Negative = ((df['y']==1) & (df['Y_pred']==0)).sum()
            
    False_Positive = ((df['y']==0) & (df['Y_pred']==1)).sum()
            
    return True_Positive ,True_Negative ,False_Negative,False_Positive






# Calculate  F1 SCORE

In [5]:
def compute_f1_score(TP, FP, FN):
    
    precision = TP / (TP + FP)
    recall = TP / (TP + FN)
    f1_score = (2*precision*recall)/(precision + recall)
    return precision, recall, f1_score


# Calculate Accuracy Score

In [6]:
def accuracy_score(TP, TN, FP, FN):
    accu_score = (TP + TN) / (TP + TN + FP + FN)
    print(accu_score,"score")
    return accu_score

# Calculate Auc Score

In [7]:
def auc_score(df):
    tpr = []
    fpr = []
    for threshold in (df["proba"].unique()):
        df['y_hat']  = np.where(df["proba"]>=threshold,1,0)
        True_Positive= ((df['y']==1) & (df['y_hat']==1)).sum()
            
        True_Negative = ((df['y']==0) & (df['y_hat']==0)).sum()
            
        False_Negative = ((df['y']==1) & (df['y_hat']==0)).sum()
            
        False_Positive = ((df['y']==0) & (df['y_hat']==1)).sum()
        tpr.append(True_Positive/(True_Positive+False_Negative))
        fpr.append(False_Positive/(False_Positive+True_Negative))
    return tpr,fpr    




In [8]:
df = add_column(df)

# Printing confusion matrix

In [9]:
TP,TN,FN,FP=confusion_matrix(df)
print("CONFUSION MATRIX**********************************************************************************")
print(TP,"True Positive")
print(TN,"True Negative")
print(FN,"False Negative")
print(FP,"False Positive")

CONFUSION MATRIX**********************************************************************************
10000 True Positive
0 True Negative
0 False Negative
100 False Positive


# Printing precision, recall, f1_score

In [10]:
precision, recall, f1_score=compute_f1_score(TP, FP, FN)
print("Precision:", precision,"Recall:", recall, "F1 Score :", f1_score)

Precision: 0.9900990099009901 Recall: 1.0 F1 Score : 0.9950248756218906


# Printing AUC Score 

In [11]:
df.sort_values("proba", inplace=True,ascending = False)
tpr_array , fpr_array = auc_score(df)
auc=np.trapz(np.asarray(tpr_array) , np.asarray(fpr_array))
print(auc,"auc score ")

0.48829900000000004 auc score 


# Printing accuracy score 

In [12]:
accuracy_score_1 = accuracy_score(TP, TN, FP, FN)
print(accuracy_score_1)

0.9900990099009901 score
0.9900990099009901


# Task for 5_b.csv

In [13]:
df=pd.read_csv('5_b.csv')
df.head(5)

,y,proba
0,0.0,0.281035
1,0.0,0.465152
2,0.0,0.352793
3,0.0,0.157818
4,0.0,0.276648


In [14]:

dataframe = add_column(df)


# Printing confusion matrix 

In [15]:

TP,TN,FN,FP=confusion_matrix(dataframe)
print("CONFUSION MATRIX**********************************************************************************")
print(TP,"True Positive")
print(TN,"True Negative")
print(FN,"False Negative")
print(FP,"False Positive")

CONFUSION MATRIX**********************************************************************************
55 True Positive
9761 True Negative
45 False Negative
239 False Positive


# Printing AUC Score 

In [16]:
precision, recall, f1_score=compute_f1_score(TP, FP, FN)

print("Precision:", precision,"Recall:", recall, "F1 Score :", f1_score)

Precision: 0.1870748299319728 Recall: 0.55 F1 Score : 0.2791878172588833


# Printing AUC Score 

In [17]:
df.sort_values("proba", inplace=True,ascending = False)
tpr_array , fpr_array = auc_score(df)
auc=np.trapz(np.asarray(tpr_array) , np.asarray(fpr_array))
print(auc)


0.9377570000000001


# Printing accuracy score

In [18]:
accu_score = accuracy_score(TP, TN, FP, FN)
print(accu_score)

0.9718811881188119 score
0.9718811881188119


<pre>
<font color='red'><b>B.</b></font> Compute performance metrics for the given data <strong>5_b.csv</strong>
   <b>Note 1:</b> in this data you can see number of positive points << number of negatives points
   <b>Note 2:</b> use pandas or numpy to read the data from <b>5_b.csv</b>
   <b>Note 3:</b> you need to derive the class labels from given score</pre> $y^{pred}= \text{[0 if y_score < 0.5 else 1]}$

<pre>
<ol>
<li> Compute Confusion Matrix </li>
<li> Compute F1 Score </li>
<li> Compute AUC Score, you need to compute different thresholds and for each threshold compute tpr,fpr and then use               numpy.trapz(tpr_array, fpr_array) <a href='https://stackoverflow.com/q/53603376/4084039'>https://stackoverflow.com/q/53603376/4084039</a>, <a href='https://stackoverflow.com/a/39678975/4084039'>https://stackoverflow.com/a/39678975/4084039</a></li>
<li> Compute Accuracy Score </li>
</ol>
</pre>

<font color='red'><b>C.</b></font> Compute the best threshold (similarly to ROC curve computation) of probability which gives lowest values of metric <b>A</b> for the given data <strong>5_c.csv</strong>
<br>

you will be predicting label of a data points like this: $y^{pred}= \text{[0 if y_score < threshold  else 1]}$

$ A = 500 \times \text{number of false negative} + 100 \times \text{numebr of false positive}$

<pre>
   <b>Note 1:</b> in this data you can see number of negative points > number of positive points
   <b>Note 2:</b> use pandas or numpy to read the data from <b>5_c.csv</b>
</pre>

In [19]:
df = pd.read_csv('5_c.csv')
thresholds = np.sort(df.prob.unique())
print(len(thresholds))

A = []

for threshold in thresholds:
    thresh_detail = {}
    df['Y_pred'] = np.where(df['prob'] >= threshold, 1, 0)
    False_Negative = ((df['Y_pred'] == 1) & (df['Y_pred'] == 0)).sum()
    False_Positive = ((df['Y_pred'] == 1) & (df['Y_pred'] == 1)).sum()

    a = 500 * False_Negative + 100 * False_Positive
    A.append(a)
print("Thresholds ", thresholds)
print("****************************************************************************************************************")
print("A ",A)

2791
Thresholds  [0.02803799 0.02839574 0.02896366 ... 0.94863779 0.95143692 0.9577468 ]
****************************************************************************************************************
A  [285200, 285100, 285000, 284900, 284800, 284700, 284600, 284500, 284400, 284300, 284200, 284100, 284000, 283900, 283800, 283700, 283600, 283500, 283400, 283300, 283200, 283100, 283000, 282900, 282800, 282700, 282600, 282500, 282400, 282300, 282200, 282100, 282000, 281900, 281800, 281700, 281600, 281500, 281400, 281300, 281200, 281100, 281000, 280900, 280800, 280700, 280600, 280500, 280400, 280300, 280200, 280100, 280000, 279900, 279800, 279700, 279600, 279400, 279300, 279200, 279100, 279000, 278900, 278800, 278700, 278600, 278500, 278400, 278300, 278200, 278100, 278000, 277900, 277800, 277700, 277600, 277500, 277400, 277300, 277200, 277100, 277000, 276900, 276800, 276700, 276600, 276500, 276400, 276300, 276200, 276100, 276000, 275900, 275800, 275700, 275600, 275500, 275400, 275300, 27

The lowest value of a is 100 whhich comes the largest value of thresold
so the best thresold value is 0.9577468

<pre>
<font color='red'><b>D.</b></font> Compute performance metrics(for regression) for the given data <strong>5_d.csv</strong>
    <b>Note 2:</b> use pandas or numpy to read the data from <b>5_d.csv</b>
    <b>Note 1:</b> <b>5_d.csv</b> will having two columns Y and predicted_Y both are real valued features
<ol>
<li> Compute Mean Square Error </li>
<li> Compute MAPE: https://www.youtube.com/watch?v=ly6ztgIkUxk</li>
<li> Compute R^2 error: https://en.wikipedia.org/wiki/Coefficient_of_determination#Definitions </li>
</ol>
</pre>

# Calculating MSE

In [20]:
df_D=pd.read_csv('5_d.csv')
count_of_y=list(df_D['y'])
count_of_pred=list(df_D['pred'])
sumtion = 0  
n = len(count_of_y) 
for i in range (0,n):  
    diff = count_of_y[i] - count_of_pred[i]  
    sq_diff = diff**2  
    sumtion = sumtion + sq_diff  
    MSE = sumtion/n  
print("The Mean Square Error is: " , MSE)

The Mean Square Error is:  177.16569974554707


# Calculating MAPE

In [21]:
y_matrix=df_D['y']
pred_matrix=df_D['pred']
def mean_abs_percentage_error(y_matrix, pred_matrix): 
    y, pred_y = np.array(y_matrix), np.array(pred_matrix)
    return np.mean(np.abs((y - pred_y) / np.mean(y)))

mean_abs_percentage_error(y_matrix, pred_matrix)

0.12912029940096867

# Calculating R Square Error

In [22]:
def RSqr(y_matrix, pred_matrix):
    sse_val = sum((y_matrix - pred_matrix)**2)
    tse_val = (len(y_matrix) - 1) * np.var(y_matrix, ddof=1)
    score_of_r2 = 1 - (sse_val / tse_val)
    return score_of_r2

r_sqr_error = RSqr(y_matrix, pred_matrix)
print(r_sqr_error)

0.9563582786990964
